[RWKV] Add RWKV5 model and RWKVWorldTokenizer

BBuf opened the pull request:
Add RWKVWorldTokenizer for rwkv5 series model.

The tokenizer has been used in:

- [RWKV/rwkv-5-world-1b5](https://huggingface.co/RWKV/rwkv-5-world-1b5)
- [RWKV/rwkv-5-world-3b](https://huggingface.co/RWKV/rwkv-5-world-3b)
- [RWKV/rwkv-4-world-169m](https://huggingface.co/RWKV/rwkv-4-world-169m)
- [RWKV/rwkv-4-world-430m](https://huggingface.co/RWKV/rwkv-4-world-430m)
- [RWKV/rwkv-4-world-1b5](https://huggingface.co/RWKV/rwkv-4-world-1b5)
- [RWKV/rwkv-4-world-3b](https://huggingface.co/RWKV/rwkv-4-world-3b)
- [RWKV/rwkv-4-world-7b](https://huggingface.co/RWKV/rwkv-4-world-7b)

and lambda test in https://github.com/BBuf/RWKV-World-HF-Tokenizer/blob/main/check_lambda/lambda_hf.py

@xianbaoqian

BBuf committed:
add rwkv world tokenizer

BBuf committed:
update init

BBuf committed:
update init

BBuf committed:
update init

BBuf committed:
refine

BBuf committed:
refine

ArthurZucker commented:
Hey feel free to ping me when this is ready! 🤗

BBuf committed:
fix ci error

BBuf committed:
fix ci error

BBuf committed:
Merge branch 'main' into main

BBuf committed:
fix ci error

BBuf committed:
fix ci error

BBuf committed:
fix ci error

BBuf committed:
fix ci error

BBuf committed:
try to fix ci error

BBuf committed:
fix dummy_pt_objects bug

BBuf committed:
fix ci error

BBuf committed:
fix requires_torch

BBuf committed:
delete useless import

BBuf committed:
refine

BBuf commented:
Hi, pr ready now  🤗. @ArthurZucker

ArthurZucker commented:
Ok! Thanks, I'll review now, but will let @amyeroberts handle the rest as I'll be off for a week 😉

ArthurZucker commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -328,7 +328,7 @@
             ("roc_bert", ("RoCBertTokenizer", None)),
             ("roformer", ("RoFormerTokenizer", "RoFormerTokenizerFast" if is_tokenizers_available() else None)),
-            ("rwkv", (None, "GPTNeoXTokenizerFast" if is_tokenizers_available() else None)),
+            ("rwkv", ("RWKVWorldTokenizer", None)),
```
We can't really do this 😓 If one is supposed to be the "slow" we can just keep both. If they are fundamentally different we need something else.

BBuf replied:
Alright, I'll fix it.

BBuf replied:
You might be right, the model code of the recently implemented version for RWKV World model has changed compared to this one. I'll consider using a new folder.

ArthurZucker commented on tests/fixtures/rwkv_vocab_v20230424.json:
this should rather be on the hub 😉

ArthurZucker commented on src/transformers/models/rwkv/tokenization_rwkv.py:
```diff
@@ -0,0 +1,525 @@
+
+    def __init__(self, vocab_file, errors="replace", **kwargs):
+        self.add_bos_token = False
+        self._bos_token = None
+        self._eos_token = None
+        self._unk_token = None
+        self._sep_token = None
+        self._pad_token = None
+        self._cls_token = None
+        self._mask_token = None
+        self._additional_special_tokens = []
```
this is done by default so not needed here

ArthurZucker commented on src/transformers/models/rwkv/tokenization_rwkv.py:
```diff
@@ -0,0 +1,525 @@
+        return res_str
+
+
+class RWKVWorldTokenizer(PreTrainedTokenizer):
+    """
+    Construct a RWKVWorldTokenizer. Based on byte-level Byte-Pair-Encoding.
```
We have a lot of BPE based tokenizers like GPT2 for example, would it work to use them?

BBuf replied:
For the RWKV World Tokenizer, it terminates at the inferred token id=0 (https://github.com/huggingface/transformers/pull/26963/files#diff-6b1cfe4e721b792b5bdfde6f7a2645a9a92fcf7b9e3fb0e8c81f040759a7b6d0R204), and there is no explicit eos (end of sentence) token in the vocabulary. This is the main source of difference.

ArthurZucker replied:
Then we should just manually set the token 0 to be the eos token. The content does not really matter you can choose whatever no?

ArthurZucker commented on src/transformers/models/rwkv/tokenization_rwkv.py:
```diff
@@ -0,0 +1,525 @@
+            elif rid == 0:
+                break
+            else:
+                print("ERROR unknown id %d" % rid)
```
let'snot have print, we should either log a warning or raise an error or do nothing

ArthurZucker commented on src/transformers/models/rwkv/tokenization_rwkv.py:
```diff
@@ -0,0 +1,525 @@
+            w = word[idx]
+            is_leaf = idx == (len(word) - 1)
+            leaf_data = data if is_leaf else None
+            # 不存在则插入
```
could we have this in english?

ArthurZucker reviewed (commented):
Thanks for the PR! Could you explain the motivation behind not using the fast tokenizer, and whether this tokenizer / slow implem of GPT2 for example. 

Mostly, this should need a new folder as it's a new model ! If we use the GPT2Tokenizer implementation then we can also just add a .md file ( like we did for flan T5 for example)

BBuf commented:
> Thanks for the PR! Could you explain the motivation behind not using the fast tokenizer, and whether this tokenizer / slow implem of GPT2 for example.
> 
> Mostly, this should need a new folder as it's a new model ! If we use the GPT2Tokenizer implementation then we can also just add a .md file ( like we did for flan T5 for example)

The model implementation is the same, only the tokenizer has different options. The tokenizer implemented in this PR is for the RWKV4 World model, but the implementation of the RWKV4 World model is exactly the same as the existing RWKV model implementation.

BBuf commented on docs/source/en/model_doc/rwkv.md:
```diff
@@ -73,6 +73,9 @@ output = model.generate(inputs["input_ids"], max_new_tokens=64, stopping_criteri
 
 [[autodoc]] RwkvConfig
 
+## RWKVWorldTokenizer
+
+[[autodoc]] RWKVWorldTokenizer
```
model is same, only tokenizer has difference.

BBuf committed:
add rwkv5 model

BBuf committed:
add rwkv5 model and run test success

BBuf committed:
revert change

BBuf committed:
revert change

BBuf committed:
fix ci error

BBuf committed:
format

BBuf committed:
fix ci error

BBuf committed:
fix ci bug

BBuf committed:
fix ci error

BBuf committed:
Merge branch 'main' into main

BBuf committed:
revert extra change

BBuf committed:
Merge branch 'main' of github.com:BBuf/transformers

BBuf committed:
revert extra change

BBuf committed:
revert extra change

BBuf committed:
fix ci bug

BBuf commented:
@ArthurZucker Hello, I have implemented the RWKV5 model and the RWKVWorldTokenizer it requires. Please review again. Thank you.

ArthurZucker commented on docs/source/en/model_doc/rwkv5.md:
```diff
@@ -0,0 +1,136 @@
+
+The RWKV5 model was proposed in [this repo](https://github.com/BlinkDL/RWKV-LM)
+
+It suggests a tweak in the traditional Transformer attention to make it linear. This way, the model can be used as recurrent network: passing inputs for timestamp 0 and timestamp 1 together is the same as passing inputs at timestamp 0, then inputs at timestamp 1 along with the state of timestamp 0 (see example below).
```
Since it uses the same modelling code I think we would rather specify the changes, like the dataset it was trained on etc.

ArthurZucker commented on docs/source/en/model_doc/rwkv5.md:
```diff
@@ -0,0 +1,136 @@
+
+This can be more efficient than a regular Transformer and can deal with sentence of any length (even if the model uses a fixed context length for training).
+
+This model was contributed by [sgugger](https://huggingface.co/sgugger).
```
```suggestion
```

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
No no, if the modelling is the same, we don't need this. It should use the AutoModel api similarly to Flan-t5 family models.

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
Pretty sure we can use Gpt2 as I mentionned, and only need to update the auto mapping of the original RWKV tokenization mapping to have this one if tokenizers is not installed. COrrect me if I am wrong but we should have the same results with both tokenizers no?

ArthurZucker commented on tests/fixtures/rwkv_vocab_v20230424.json:
Let's remove this

ArthurZucker reviewed (commented):
I was off for a week sorry for the delay

BBuf committed:
Merge branch 'main' into main

ArthurZucker commented:
Okay, understood! So this new model uses a word level tokenizer, which can be supported both in transformers (by adding a new tokenizer, with a simple vocab / the code you are proposing) and `tokenizers` which natively has a [`WordLevel`](https://huggingface.co/docs/tokenizers/api/tokenizer#tokenizers.models.WordLevel) tokenizer ! 

I'm thrilled to help you get this merged 😉

BBuf committed:
support batch inference for rwkv5

BBuf committed:
Merge branch 'main' of github.com:BBuf/transformers

BBuf committed:
code format

BBuf committed:
fix doc

BBuf committed:
fix ci error

BBuf committed:
fix doc error

BBuf commented:
> Okay, understood! So this new model uses a word level tokenizer, which can be supported both in transformers (by adding a new tokenizer, with a simple vocab / the code you are proposing) and `tokenizers` which natively has a [`WordLevel`](https://huggingface.co/docs/tokenizers/api/tokenizer#tokenizers.models.WordLevel) tokenizer !
> 
> I'm thrilled to help you get this merged 😉

Hello, could you take another look at this PR? The recent few commits have added support for batch inference, and I feel it's getting close to being merged.

BBuf committed:
fix tokenizer bug

BBuf committed:
refine

ArthurZucker commented:
Sure I’ll review today! 🤗

BBuf committed:
add pad_token_id

BBuf committed:
refine

BBuf committed:
format

BBuf committed:
add cuda kernel

BBuf committed:
fix format

BBuf committed:
fix bug

PicoCreator commented:
> Okay, understood! So this new model uses a word level tokenizer, which can be supported both in transformers (by adding a new tokenizer, with a simple vocab / the code you are proposing) and `tokenizers` which natively has a [`WordLevel`](https://huggingface.co/docs/tokenizers/api/tokenizer#tokenizers.models.WordLevel) tokenizer !
> 
> I'm thrilled to help you get this merged 😉

I would not call it a "word level" more of a "trie tokenizer", spaces are just simply another character with no special meaning - If that makes sense.

But yes, that in concept this tokenizer could be used for non RWKV architecture, and there is nothing stopping anyone from using our older GPT-neox tokenizer on our newer architecture.

Do let me know if I can clarify anything else from our end, or help in this merge =)

ArthurZucker commented:
Okay! I'll let you know, sorry I got caught up in sprints here and there but will review this early next week 🤗

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+}
+
+
+class TRIE:
```
let's use camel and let's try to make it a little bit more verbose with more meaningful variables and documentation. There is a trie in https://github.com/huggingface/transformers/blob/18cbaf13dcaca7145f5652aefb9b19734c56c3cd/src/transformers/tokenization_utils.py#L51-L57 

would be nice to know the differences (why is this on specifically needed)

BBuf replied:
ok, I will explain at the begining.

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+            idx = int(l[: l.index(" ")])
+            x = eval(l[l.index(" ") : l.rindex(" ")])
+            x = x.encode("utf-8") if isinstance(x, str) else x
+            assert isinstance(x, bytes)
+            assert len(x) == int(l[l.rindex(" ") :])
```
No asserts in the codebase

BBuf replied:
ok

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+            **kwargs,
+        )
+
+    @property
+    def eos_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def eot_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def pad_token_id(self) -> Optional[int]:
+        return 0
+
```
these should not be needed and should rely on the special tokens for eos and pad

BBuf replied:
In RWKV tokenizer, the token id for bos, eos, and pad are all 0. This is not implemented by adding special tokens, but rather by directly rewriting the relevant attributes, because eos_token_id and pad_token_id are already being used by default in the decoding and batch encoding func of this tokenizer.

ArthurZucker replied:
Alright, this can be done with adding the content of the first token and init with eos_token = pad_token (the super class will have this information) but can be solved afterwards no worries it's minimal !

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+    def get_vocab(self):
+        return dict(self.encoder, **self.added_tokens_encoder)
+
+    def add_tokens(self, new_tokens, special_tokens: bool = False):
+        for token in new_tokens:
+            token_id = self.convert_tokens_to_ids(token)
+            self.added_tokens_decoder[token_id] = token
```
Is this to make it **faster** than using the original logic? The issue is that this does not seperate between the original tokens and the added tokens which are usually handled differently. This should be `add_tokens_to_vocab` to not change the API for our users ! 🤗

BBuf replied:
Ok, that's right.

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+        skip_special_tokens: bool = False,
+        **kwargs,
+    ) -> str:
+        def remove_zeros_from_first_segment(token_ids, first_max_length):
```
let's not define this if we only us it once

BBuf replied:
done

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+    ):
+        return PaddingStrategy.LONGEST, TruncationStrategy.DO_NOT_TRUNCATE, -1, kwargs
+
+    def _encode_plus(
```
you should not have to overwrite this if the input is just text. Tap into the `tokenize` function rather than this

BBuf replied:
The reason for overloading `_encode_plus` and `_batch_encode_plus` is that I have to manually pad and call the custom decode implementation of RWKVWorldHFTokenizer. This custom decoder implementation handles non-utf-8 encoded token decoding. If it is not overloaded, some exceptions will be triggered, such as:

<img width="441" alt="图片" src="https://github.com/huggingface/transformers/assets/35585791/6ecf6150-9df7-4426-b03f-a9d7c748a188">

ArthurZucker replied:
Could you give the prompt used in this case? 🤗

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+            verbose=verbose,
+        )
+
+    def _batch_encode_plus(
```
same here

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
```diff
@@ -0,0 +1,545 @@
+
+        return BatchEncoding(batch_outputs)
+
+    def decode(
```
no need to overwrite no?

BBuf replied:
yes! done now.

ArthurZucker commented on tests/models/rwkv5/test_tokenization_rwkv5.py:
```diff
@@ -0,0 +1,63 @@
+
+
+@require_torch
+class RWKVWorldTokenizationTest(unittest.TestCase):
```
let's make sure the tokenizer is fully tested with the TokenizerTesterMixin

BBuf replied:
Because the RWKVWorldHFTokenizer has some special settings, such as eos_id=bos_id=0, many tests fail when using the TokenizerTesterMixin to test the tokenizer. Therefore, can we not use the TokenizerTesterMixin for testing in this case?

ArthurZucker replied:
No I think the way to fix this is to properly set eos_token and bos_token no?

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,856 @@
+rwkv5_cuda_kernel = None
+
+
+def load_wkv5_cuda_kernel(head_size):
```
This is `# Copied from` https://github.com/huggingface/transformers/blob/18cbaf13dcaca7145f5652aefb9b19734c56c3cd/src/transformers/models/rwkv/modeling_rwkv.py#L65 
with rwkv->rwkv5

BBuf replied:
done.

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,856 @@
+    rwkv5_cuda_kernel.head_size = head_size
+
+
+class WKV_5(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, B, T, C, H, r, k, v, w, u, s):
+        with torch.no_grad():
+            assert r.dtype == torch.bfloat16
+            assert k.dtype == torch.bfloat16
+            assert v.dtype == torch.bfloat16
+            assert w.dtype == torch.bfloat16
+            assert u.dtype == torch.bfloat16
+            assert s.dtype == torch.float32
+            ctx.B = B
+            ctx.T = T
+            ctx.C = C
+            ctx.H = H
+            assert r.is_contiguous()
+            assert k.is_contiguous()
+            assert v.is_contiguous()
+            assert w.is_contiguous()
+            assert u.is_contiguous()
+            ew = (-torch.exp(w.float())).contiguous()
+            eew = (torch.exp(ew)).contiguous()
+            ctx.save_for_backward(r, k, v, eew, ew, u)
+            y = torch.empty(
+                (B, T, C), device=r.device, dtype=torch.bfloat16, memory_format=torch.contiguous_format
+            )  # .uniform_(-1, 1)
+            rwkv5_cuda_kernel.forward(B, T, C, H, r, k, v, eew, u, y, s)
+            return y, s
+
```
we don't use assert in the modeling code, and let's use more meaningful variable names no single letter ones. Also why do we not support gradients ?

BBuf replied:
Done! backward supported now.

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,856 @@
+            return (None, None, None, None, gr, gk, gv, gw, gu)
+
+
+def rwkv_linear_attention_v5_cpu(
```
Let's try to follow namings from 
https://github.com/huggingface/transformers/blob/18cbaf13dcaca7145f5652aefb9b19734c56c3cd/src/transformers/models/rwkv/modeling_rwkv.py#L203

ArthurZucker replied:
Also we try to not always pas as many args for such a fonction, shapes could be passed as a tuple, etc

BBuf replied:
ok

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,856 @@
+    return out, state
+
+
+def rwkv_linear_attention(
```
same here would be good if we have a similar api as https://github.com/huggingface/transformers/blob/18cbaf13dcaca7145f5652aefb9b19734c56c3cd/src/transformers/models/rwkv/modeling_rwkv.py#L247

BBuf replied:
done.

ArthurZucker reviewed (commented):
Thanks a lot! 🤗 Let's try to make the code as clean as possible, but don't hesitate to explain your choices as well, sorry for my lack of knowledge

BBuf commented on docs/source/en/model_doc/rwkv5.md:
```diff
@@ -0,0 +1,137 @@
+outputs = model(inputs["input_ids"][:, 2:], state=outputs.state)
+output_two = outputs.last_hidden_state
+
+self.parent.assertTrue(torch.allclose(output_two, output_whole, atol=1e-5))
```
ok.

BBuf commented on src/transformers/models/rwkv5/configuration_rwkv5.py:
```diff
@@ -0,0 +1,120 @@
+
+logger = logging.get_logger(__name__)
+
+RWKV5_PRETRAINED_CONFIG_ARCHIVE_MAP = {}
```
Okay, I will fix it.

BBuf commented on src/transformers/models/rwkv5/configuration_rwkv5.py:
```diff
@@ -0,0 +1,120 @@
+RWKV5_PRETRAINED_CONFIG_ARCHIVE_MAP = {}
+
+
+class Rwkv5Config(PretrainedConfig):
```
No, it involves adding the `num_attention_heads`, `attention_hidden_size`

BBuf commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,691 @@
+    lxb,
+    ow,
+    state,
+):
```
Okay, I will fix it.

BBuf commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,691 @@
+        self.gate = nn.Linear(hidden_size, attention_hidden_size, bias=False)
+        self.output = nn.Linear(attention_hidden_size, hidden_size, bias=False)
+        # https://github.com/BlinkDL/RWKV-LM/blob/3db37a72356b736966ddd377268f02b80963af3f/RWKV-v4neo/src/model.py#L190C1-L190C1
+        self.ln_x = nn.GroupNorm(hidden_size // config.head_size, hidden_size)
```
In the original checkpoint, the weight of this group norm is saved in the dictionary with ln_x as the key, so it has not been modified here.

ArthurZucker replied:
alright!

BBuf commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,691 @@
+            # https://github.com/BlinkDL/ChatRWKV/blob/main/rwkv_pip_package/src/rwkv/model.py#L904-L906
+            state = []
+            num_attention_heads = self.config.hidden_size // self.config.num_attention_heads
+            state.append(
+                torch.zeros(
+                    (inputs_embeds.size(0), self.config.hidden_size, self.config.num_hidden_layers),
+                    dtype=inputs_embeds.dtype,
+                    requires_grad=False,
+                    device=inputs_embeds.device,
+                ).contiguous()
+            )
+            state.append(
+                torch.zeros(
+                    (
+                        inputs_embeds.size(0),
+                        num_attention_heads,
+                        self.config.hidden_size // num_attention_heads,
+                        self.config.hidden_size // num_attention_heads,
+                        self.config.num_hidden_layers,
+                    ),
+                    dtype=torch.float32,
+                    requires_grad=False,
+                    device=inputs_embeds.device,
+                ).contiguous()
+            )
+            state.append(
+                torch.zeros(
+                    (inputs_embeds.size(0), self.config.hidden_size, self.config.num_hidden_layers),
+                    dtype=inputs_embeds.dtype,
+                    requires_grad=False,
+                    device=inputs_embeds.device,
+                ).contiguous()
```
good comment.

BBuf commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,691 @@
+                        block.feed_forward.value.weight.div_(2 ** int(block_id // self.config.rescale_every))
+
+        self.layers_are_rescaled = not self.training
+
```
ok, I will add it.

BBuf committed:
fix part comment

BBuf committed:
Merge branch 'main' into main

BBuf committed:
Merge branch 'huggingface:main' into main

BBuf committed:
fix more comment

BBuf committed:
delete useless decode overwrite

BBuf committed:
fix comment

BBuf committed:
fix bug

ArthurZucker commented:
All the progress look good! Ping me whenever for another review! 🤗

winglian commented:
Now that rwkv5 pretrained model is out, will this get merged?

ArthurZucker commented:
I'll review again and help merge it asap!

BBuf committed:
Merge branch 'main' into main

BBuf committed:
try to fix ci

BBuf committed:
format

ArthurZucker commented on src/transformers/models/rwkv5/tokenization_rwkv5.py:
The issue is mostly that the `tokenize` function for example does not return the tokens. Most of the api is tested in the `TokenizerTestMixin` which is why it is so important!

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,894 @@
+# coding=utf-8
+# Copyright 2023 Bo Peng and HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
```suggestion
# Copyright 2024 Bo Peng and HuggingFace Inc. team.
```

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,894 @@
+        return receptance * value, state
+
+
+# copied from HuggingFace https://github.com/huggingface/transformers/blob/main/src/transformers/models/rwkv/modeling_rwkv.py
```
```suggestion
# Copied from transformers.models.rwkv.modeling_rwkv.RwkvBlock with Rwkv->Rwkv5
```

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,894 @@
+rwkv5_cuda_kernel = None
+
+
+# Copied from https://github.com/huggingface/transformers/blob/18cbaf13dcaca7145f5652aefb9b19734c56c3cd/src/transformers/models/rwkv/modeling_rwkv.py#L65
```
```suggestion
# Copied from transformers.models.rwkv.modeling_rwkv.load_wkv_cuda_kernel with wkv->wkv5
```

ArthurZucker commented on src/transformers/models/rwkv5/modeling_rwkv5.py:
```diff
@@ -0,0 +1,894 @@
+                module.time_mix_receptance.data = torch.pow(time_weight, ratio_1_to_almost0)
+
+
+# copied from HuggingFace https://github.com/huggingface/transformers/blob/main/src/transformers/models/rwkv/modeling_rwkv.py
```
```suggestion
# Copied from transformers.models.rwkv.modeling_rwkv.RwkvOutput with Rwkv->Rwkv5
```

ArthurZucker reviewed (commented):
Great work! 
- is there a reason why only inference is supported while we have the backwards kernel impl? 
- `# Copied from transformers.models.rwkv.modeling_rwkv.load_wkv_cuda_kernel with wkv->wkv5` should help you with what pieces that are basically almost duplicate from the previous implem. `make fix-copies` will overwrite accordingly
- regarding the tokenizer, I still think we can handle bytes easily, will try something 
- modeling, to make it compatible with the `transformers` library philosophy but also because this model is so important we need to make sure we deliver a clean code! 
For that I think the RWKV modeling code is a good start, it eliminated most of the args in each function , keeping the most essential like `def forward(ctx, time_decay, time_first, key, value, state=None, return_state=False)`. 
- we need to respect conventions, camel is used for classes, while batch should be lower case, `batch_size` 

Other than that, seems in a very good state! We need a bit more tokenizer test but I can help you with that!

ArthurZucker reviewed (commented):
The closest tokenizer is the `bert` tokenizer, which uses a WordPiece tokenizer:

https://github.com/huggingface/transformers/blob/c50602bbe9713308bf02d905535e45429f2bfb08/src/transformers/models/bert/tokenization_bert.py#L534

Here is a more transformers like version of the tokenizer: 
```python3
# coding=utf-8
# Copyright 2024 The HuggingFace Inc. team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tokenization classes for RWKV5."""

import os
from typing import TYPE_CHECKING, List, Optional, Tuple, Union

from transformers.tokenization_utils import PreTrainedTokenizer, AddedToken
from transformers.tokenization_utils_base import (
    BatchEncoding,
    EncodedInput,
    TextInput,
    TruncationStrategy,
)
from transformers.utils import PaddingStrategy, TensorType, logging, to_py_obj


if TYPE_CHECKING:
    pass

logger = logging.get_logger(__name__)

VOCAB_FILES_NAMES = {
    "vocab_file": "rwkv_vocab_v20230424.txt",
}
PRETRAINED_VOCAB_FILES_MAP = {
    "vocab_file": {
        "RWKV/rwkv-5-world-1b5": "https://huggingface.co/RWKV/rwkv-5-world-1b5/blob/main/rwkv_vocab_v20230424.txt",
    },
}

import re
def whitespace_tokenize(text):
    """Runs basic whitespace cleaning and splitting on a piece of text.
    The separators are kept
    """
    text = text.strip()
    if not text:
        return []
    tokens = re.split(r'(?=\W)', text)
    return tokens


class WordpieceTokenizer(object):
    """Runs WordPiece tokenization."""

    def __init__(self, vocab, unk_token, max_input_chars_per_word=100):
        self.vocab = vocab
        self.unk_token = unk_token
        self.max_input_chars_per_word = max_input_chars_per_word


    def tokenize(self, text):
        """
        Tokenizes a piece of text into its word pieces. This uses a greedy longest-match-first algorithm to perform
        tokenization using the given vocabulary.

        For example, `input = "unaffable"` wil return as output `["un", "##aff", "##able"]`.

        Args:
            text: A single token or whitespace separated tokens. This should have
                already been passed through *BasicTokenizer*.

        Returns:
            A list of wordpiece tokens.
        """

        output_tokens = []
        for token in whitespace_tokenize(text):
            chars = list(token)
            if len(chars) > self.max_input_chars_per_word:
                output_tokens.append(self.unk_token)
                continue

            is_bad = False
            start = 0
            sub_tokens = []
            while start < len(chars):
                end = len(chars)
                cur_substr = None
                while start < end:
                    substr = "".join(chars[start:end])
                    if substr.encode('utf-8') in self.vocab:
                        cur_substr = substr
                        break
                    end -= 1
                if cur_substr is None:
                    is_bad = True
                    break
                sub_tokens.append(cur_substr)
                start = end

            if is_bad:
                output_tokens.append(self.unk_token)
            else:
                output_tokens.extend(sub_tokens)
        return output_tokens

class Rwkv5Tokenizer(PreTrainedTokenizer):
    
    def __init__(self, vocab_file,bos_token="<s>", eos_token="<s>",unk_token="<s>",**kwargs):
        if not os.path.isfile(vocab_file):
            raise ValueError(
                f"Can't find a vocabulary file at path '{vocab_file}'. To load the vocabulary from a Google pretrained"
                " model use `tokenizer = BertTokenizer.from_pretrained(PRETRAINED_MODEL_NAME)`"
            )
            
        with open(vocab_file, "rb") as reader:
            tokens = reader.readlines()
        vocab = {}
        for index, token in enumerate(tokens):
            token = token.rstrip(b"\n")
            vocab[token] = index

        self.add_bos_token = True
        self.encoder = vocab
        self.decoder = {v:k for k,v in vocab.items()}
        self.wordpiece_tokenizer = WordpieceTokenizer(vocab=self.encoder, unk_token=str(unk_token))
        self._added_tokens_decoder = {0:AddedToken(bos_token)}
        super().__init__(
            bos_token=bos_token,
            eos_token=eos_token,
            unk_token=unk_token,
            **kwargs
        )

    @property
    def vocab_size(self):
        return len(self.encoder)

    def get_vocab(self):
        return dict(self.encoder, **self.added_tokens_encoder)
    
    def _tokenize(self, text, split_special_tokens=False):
        return self.wordpiece_tokenizer.tokenize(text)
    
    def _convert_token_to_id(self, token):
        """Converts a token (str) in an id using the vocab."""
        return self.encoder.get(token.encode("utf-8"), self.unk_token_id)

    def _convert_id_to_token(self, index):
        """Converts an index (integer) in a token (str) using the vocab."""
        return self.decoder.get(index, self.unk_token).decode("utf-8")

    def convert_tokens_to_string(self, tokens):
        """Converts a sequence of tokens (string) in a single string."""
        out_string = "".join(tokens)
        return out_string

    def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] = None) -> Tuple[str]:
        index = 0
        if os.path.isdir(save_directory):
            vocab_file = os.path.join(
                save_directory, (filename_prefix + "-" if filename_prefix else "") + VOCAB_FILES_NAMES["vocab_file"]
            )
        else:
            vocab_file = (filename_prefix + "-" if filename_prefix else "") + save_directory
        with open("rwkv.txt", "wb") as writer:
            for token, token_index in sorted(self.encoder.items(), key=lambda kv: kv[1]):
                if index != token_index:
                    logger.warning(
                        f"Saving vocabulary to {vocab_file}: vocabulary indices are not consecutive."
                        " Please check that the vocabulary is not corrupted!"
                    )
                    index = token_index
                writer.write(token + b"\n")
                index += 1
        return (vocab_file,)
    
    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
        if self.add_bos_token:
            bos_token_ids = [self.bos_token_id]
        else:
            bos_token_ids = []

        output = bos_token_ids + token_ids_0

        if token_ids_1 is None:
            return output

        return output + bos_token_ids + token_ids_1

    def get_special_tokens_mask(
        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None, already_has_special_tokens: bool = False
    ) -> List[int]:
        """
        Retrieves sequence ids from a token list that has no special tokens added. This method is called when adding
        special tokens using the tokenizer `prepare_for_model` or `encode_plus` methods.

        Args:
            token_ids_0 (`List[int]`):
                List of IDs.
            token_ids_1 (`List[int]`, *optional*):
                Optional second list of IDs for sequence pairs.
            already_has_special_tokens (`bool`, *optional*, defaults to `False`):
                Whether or not the token list is already formatted with special tokens for the model.

        Returns:
            `List[int]`: A list of integers in the range [0, 1]: 1 for a special token, 0 for a sequence token.
        """
        if already_has_special_tokens:
            return super().get_special_tokens_mask(
                token_ids_0=token_ids_0, token_ids_1=token_ids_1, already_has_special_tokens=True
            )

        if not self.add_bos_token:
            return super().get_special_tokens_mask(
                token_ids_0=token_ids_0, token_ids_1=token_ids_1, already_has_special_tokens=False
            )

        if token_ids_1 is None:
            return [1] + ([0] * len(token_ids_0))
        return [1] + ([0] * len(token_ids_0)) + [1] + ([0] * len(token_ids_1))

```

ArthurZucker commented:
This yields the following:
```python 
from transformers import Rwkv5Tokenizer
tokenizer = Rwkv5Tokenizer("/Users/arthurzucker/Work/transformers/rwkv.txt")
prompt = "Hey how are you? 男：听说你们公司要派你去南方工作"
ids = tokenizer.encode(prompt)
[0, 6037, 21887, 21338, 22851, 64, 65517, 14631, 19181, 11095, 16765, 10494, 10432, 10708, 11059, 16533, 13848, 10494, 11015, 10964, 13066, 12167, 10490]
['Hey', ' how', ' are', ' you', '?', ' ', '男', '：', '听', '说', '你', '们', '公', '司', '要', '派', '你', '去', '南', '方', '工', '作']
['<s>', 'Hey', ' how', ' are', ' you', '?', ' ', '男', '：', '听', '说', '你', '们', '公', '司', '要', '派', '你', '去', '南', '方', '工', '作']
<s>Hey how are you? 男：听说你们公司要派你去南方工作
<s>Hey how are you? 男：听说你们公司要派你去南方工作
```

BBuf commented:
> This yields the following:
> 
> ```python
> >>> from transformers import Rwkv5Tokenizer
> >>> tokenizer = Rwkv5Tokenizer("/Users/arthurzucker/Work/transformers/rwkv.txt")
> >>> prompt = "Hey how are you? 男：听说你们公司要派你去南方工作"
> >>> ids = tokenizer.encode(prompt)
> 
> >>> print(ids)
> [0, 6037, 21887, 21338, 22851, 64, 65517, 14631, 19181, 11095, 16765, 10494, 10432, 10708, 11059, 16533, 13848, 10494, 11015, 10964, 13066, 12167, 10490]
> >>> print(tokenizer.tokenize(prompt))
> ['Hey', ' how', ' are', ' you', '?', ' ', '男', '：', '听', '说', '你', '们', '公', '司', '要', '派', '你', '去', '南', '方', '工', '作']
> >>> print(tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt)))
> ['<s>', 'Hey', ' how', ' are', ' you', '?', ' ', '男', '：', '听', '说', '你', '们', '公', '司', '要', '派', '你', '去', '南', '方', '工', '作']
> >>> print(tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt))))
> <s>Hey how are you? 男：听说你们公司要派你去南方工作
> >>> print(tokenizer.decode(tokenizer.encode(prompt)))
> <s>Hey how are you? 男：听说你们公司要派你去南方工作
> ```

Thank you for your advice. The main problem here is that the original tokenizer implementation(https://github.com/BlinkDL/ChatRWKV/tree/main/tokenizer) does not have bos, eos, or pad token, but bos_token_id, eos_token_id, and pad_token_id are all set to `0` . In my implementation on Hugging Face, I have simulated this situation. But now I am unsure what to set for bos, eos, and pad token, as it seems that setting any token would not meet expectations. Therefore, it feels like this tokenizer is a special hack case. I would like to ask if it is acceptable for the tokenizer's definition not to be merged into Transformers repository, and only merge the implementation of the RWKV5 model. Then, this custom tokenizer implementation could be placed in the corresponding repository on Hugging Face, for example: https://huggingface.co/RWKV/HF_v5-Eagle-7B .

ArthurZucker commented:
In the code I provided I manually set `self._added_tokens_decoder = {0:AddedToken(bos_token)}` 
which forces the token 0 to be the `bos_token`. We can of course force any other behaviour that way, but we only need to define a string token. This will be *very* useful to the community as a whole for any SFT training, as it's the expected api.

ArthurZucker commented:
Whether or not the original tokenizer has a `token`, it has `token_id=0` which means we can choose the content of the token. I used `<s>` but we should use something like `<|endoftext|>` just to make sure it doesn't exist. This should solve all the issues you are having

ArthurZucker commented:
WDYT?

PicoCreator commented:
In general the model was not trained with a special bos_token, or pad_token in mind

However for all these tokens, we typically just use token 0 as fallback, and it "generally works" if that makes sense - so i think defaulting for all these tokens as 0 makes sense to me (coming from the trainer / model team side)

LuciferianInk commented:
I ran into a bug with this tokenizer, [sourced from here](https://huggingface.co/RWKV/rwkv-5-world-1b5/blob/main/modeling_rwkv5.py). I'm not sure how much the two code bases have diverged at this point, but @BlinkDL asked me to report it in this PR.

My issue relates to this code:
```py
tokenized_batches = []
with tqdm(total=len(batches)) as pbar:
    for batch in batches:
        tokenized = tokenizer(
            batch,
            max_length=block_size,
            stride=stride,
            padding="max_length",
            return_overflowing_tokens=True,
            truncation=True,
            return_tensors="np",
        )
        tokenized_batches.append(tokenized["input_ids"])
        pbar.update(1)

tokens = np.concatenate(tokenized_batches)
```
Typically, the tokenizer should pad these batches to a consistent length, but that's not happening here:
```
  File "/usr/local/lib/python3.10/dist-packages/aigen/datasets.py", line 188, in encode_tokens
    tokens = np.concatenate(tokenized_batches)
ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 24226 and the array at index 1 has size 26680
```
The fix for me was rather simple, but I feel like it should probably be handled by the tokenizer itself:
```
padded_batches = []
for batch in tokenized_batches:
    # Calculate the number of padding tokens needed
    padding_length = max_len - batch.shape[1]
    # Pad the batch and add it to the padded_batches list
    if padding_length > 0:
        padded_batch = np.pad(batch, ((0, 0), (0, padding_length)), mode='constant', constant_values=tokenizer.pad_token_id)
    else:
        padded_batch = batch
    padded_batches.append(padded_batch)
```
Just a PSA. Thanks!

ArthurZucker commented:
Can you try with this one: https://github.com/huggingface/transformers/pull/26963#pullrequestreview-1861671950

BBuf commented:
> Can you try with this one: [#26963 (review)](https://github.com/huggingface/transformers/pull/26963#pullrequestreview-1861671950)

Thank you, I will have a try.

ArthurZucker commented:
I converted the vocab to the appropriate format to read it, sorry I forgot that step. Will push it now

BBuf commented:
> I converted the vocab to the appropriate format to read it, sorry I forgot that step. Will push it now. I used your tokenizer's `encoder`

Okay, thanks.

ArthurZucker commented:
I used this :
```python 
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("RWKV/rwkv-5-world-1b5", trust_remote_code=True)

with open("/Users/arthurzucker/Work/transformers/rwkv.txt", "wb") as f:
    for index, token in sorted(tokenizer.decoder.items(),  key=lambda kv: kv[1]):
        f.write(token +  b"\n")


tokenizer = Rwkv5Tokenizer("/Users/arthurzucker/Work/transformers/rwkv.txt")
prompt = "Hey how are you? 男：听说你们公司要派你去南方工作"
ids = tokenizer.encode(prompt)
print(ids)
print(tokenizer.tokenize(prompt))
print(tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt)))
print(tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(tokenizer.encode(prompt))))
print(tokenizer.decode(tokenizer.encode(prompt)))
```

ArthurZucker commented:
Pushed the tokenizer here: https://huggingface.co/ArthurZ/rwkv-5

BBuf commented:
> Pushed the tokenizer here: https://huggingface.co/ArthurZ/rwkv-5

Okay, got it.

BBuf committed:
try to fix tokenizer

BBuf commented:
> Pushed the tokenizer here: https://huggingface.co/ArthurZ/rwkv-5

Hello, I encountered an error while testing this tokenizer, but I'm not sure how to resolve it.

```
ERROR: test_added_token_serializable (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_added_token_serializable) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2204, in test_added_token_serializable
    tokenizer.from_pretrained(tmp_dir_name)
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils_base.py", line 2031, in from_pretrained
    return cls._from_pretrained(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils_base.py", line 2263, in _from_pretrained
    tokenizer = cls(*init_inputs, **init_kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/models/rwkv5/tokenization_rwkv5.py", line 133, in __init__
    self._added_tokens_decoder = {0:AddedToken(bos_token)}
                                    ^^^^^^^^^^^^^^^^^^^^^
TypeError: argument 'content': 'AddedToken' object cannot be converted to 'PyString'
```

ArthurZucker commented:
Could you share your testing! It should be easy to solve: 
```diff 
-self._added_tokens_decoder = {0:AddedToken(bos_token)}
+self._added_tokens_decoder = {0:AddedToken(str(bos_token))}
```
should fix it

BBuf committed:
fix bug

BBuf commented:
> Could you share your testing! It should be easy to solve:
> 
> ```diff
> -self._added_tokens_decoder = {0:AddedToken(bos_token)}
> +self._added_tokens_decoder = {0:AddedToken(str(bos_token))}
> ```
> 
> should fix it

Thank you. My test is the `test_tokenization_rwkv5.py` in this PR. Currently, some tokenizer tests have not passed, and I am trying to solve them. 

```
....Saving vocabulary to /var/folders/k6/cvg5yzcs1957ywm2m80_y3dw0000gn/T/tmpqpufnio8/rwkv_vocab_v20230424.txt: vocabulary indices are not consecutive. Please check that the vocabulary is not corrupted!
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
.....Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
...s..Saving vocabulary to /var/folders/k6/cvg5yzcs1957ywm2m80_y3dw0000gn/T/tmp_epovmhw/rwkv_vocab_v20230424.txt: vocabulary indices are not consecutive. Please check that the vocabulary is not corrupted!
.....E......E....Token indices sequence length is longer than the specified maximum sequence length for this model (2108 > 100). Running this sequence through the model will result in indexing errors
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
Be aware, overflowing tokens are not returned for the setting you have chosen, i.e. sequence pairs with the 'longest_first' truncation strategy. So the returned list will always be empty even if some tokens have been removed.
.Token indices sequence length is longer than the specified maximum sequence length for this model (2100 > 100). Running this sequence through the model will result in indexing errors
..s......Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils_base.py:2621: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  warnings.warn(
.s.s.....FF.Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.
...Saving vocabulary to /var/folders/k6/cvg5yzcs1957ywm2m80_y3dw0000gn/T/tmpt4b3xthf/rwkv_vocab_v20230424.txt: vocabulary indices are not consecutive. Please check that the vocabulary is not corrupted!
FSaving vocabulary to /var/folders/k6/cvg5yzcs1957ywm2m80_y3dw0000gn/T/tmpi80ljan7/rwkv_vocab_v20230424.txt: vocabulary indices are not consecutive. Please check that the vocabulary is not corrupted!
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
F..........Saving vocabulary to /var/folders/k6/cvg5yzcs1957ywm2m80_y3dw0000gn/T/tmpuir4d9zz/rwkv_vocab_v20230424.txt: vocabulary indices are not consecutive. Please check that the vocabulary is not corrupted!
......s........Using bos_token, but it is not set yet.
Using eos_token, but it is not set yet.
Using unk_token, but it is not set yet.
Using sep_token, but it is not set yet.
Using pad_token, but it is not set yet.
Using cls_token, but it is not set yet.
Using mask_token, but it is not set yet.
Using additional_special_tokens, but it is not set yet.
..s...
======================================================================
ERROR: test_conversion_reversible (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_conversion_reversible) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2074, in test_conversion_reversible
    self.assertEqual(tokenizer.convert_tokens_to_ids(word), ind)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils.py", line 649, in convert_tokens_to_ids
    ids.append(self._convert_token_to_id_with_added_voc(token))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils.py", line 658, in _convert_token_to_id_with_added_voc
    return self._convert_token_to_id(token)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/models/rwkv5/tokenization_rwkv5.py", line 153, in _convert_token_to_id
    return self.encoder.get(token.encode("utf-8"), self.unk_token_id)
                            ^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'encode'

======================================================================
ERROR: test_get_vocab (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_get_vocab) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2059, in test_get_vocab
    vocab = [tokenizer.convert_ids_to_tokens(i) for i in range(len(tokenizer))]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2059, in <listcomp>
    vocab = [tokenizer.convert_ids_to_tokens(i) for i in range(len(tokenizer))]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/tokenization_utils.py", line 973, in convert_ids_to_tokens
    return self._convert_id_to_token(ids)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/transformers-4.38.0.dev0-py3.11.egg/transformers/models/rwkv5/tokenization_rwkv5.py", line 157, in _convert_id_to_token
    return self.decoder.get(index, self.unk_token).decode("utf-8")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'decode'

======================================================================
FAIL: test_pretokenized_inputs (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_pretokenized_inputs) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2280, in test_pretokenized_inputs
    self.assertEqual(output, output_sequence)
AssertionError: Lists differ: [66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 7[29 chars], 85] != [66, 33, 67, 68, 69, 70, 71, 72, 73, 74, 75, 7[33 chars], 85]

First differing element 1:
67
33

Second list contains 1 additional elements.
First extra element 20:
85

- [66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85]
+ [66,
+  33,
+  67,
+  68,
+  69,
+  70,
+  71,
+  72,
+  73,
+  74,
+  75,
+  76,
+  77,
+  78,
+  79,
+  80,
+  81,
+  82,
+  83,
+  84,
+  85]

======================================================================
FAIL: test_pretrained_model_lists (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_pretrained_model_lists)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 1027, in test_pretrained_model_lists
    self.assertGreaterEqual(len(self.tokenizer_class.pretrained_vocab_files_map), 1)
AssertionError: 0 not greater than or equal to 1

======================================================================
FAIL: test_save_and_load_tokenizer (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_save_and_load_tokenizer) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 718, in test_save_and_load_tokenizer
    self.assertListEqual(before_tokens, after_tokens)
AssertionError: Lists differ: [0, 0, 0, 0, 45, 0, 0] != [0, 0, 0, 0, 44, 0, 0]

First differing element 4:
45
44

- [0, 0, 0, 0, 45, 0, 0]
?               ^

+ [0, 0, 0, 0, 44, 0, 0]
?               ^


======================================================================
FAIL: test_save_and_load_tokenizer (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_save_and_load_tokenizer) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 743, in test_save_and_load_tokenizer
    self.assertListEqual(before_tokens, after_tokens)
AssertionError: Lists differ: [0, 0, 0, 0, 45, 0, 0] != [0, 0, 0, 0, 44, 0, 0]

First differing element 4:
45
44

- [0, 0, 0, 0, 45, 0, 0]
?               ^

+ [0, 0, 0, 0, 44, 0, 0]
?               ^
```

ArthurZucker commented:
Alright I’ll checkout the PR to help out with the test !

BBuf committed:
fix test bug

BBuf committed:
fix test bug

BBuf commented:
> Alright I’ll checkout the PR to help out with the test !


Thank you, I indeed encountered difficulties while resolving the remaining tests for the tokenizer. I don't quite understand the specific reasons behind these errors. The tests that are throwing these errors are as follows:

```
======================================================================
FAIL: test_conversion_reversible (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_conversion_reversible) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 2075, in test_conversion_reversible
    self.assertEqual(tokenizer.convert_ids_to_tokens(ind), word)
AssertionError: '<s>' != b'\x00'

======================================================================
FAIL: test_internal_consistency (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_internal_consistency) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 984, in test_internal_consistency
    self.assertEqual(text_2, output_text)
AssertionError: '<s><s><s>' != 'DEF'
- <s><s><s>
+ DEF


======================================================================
FAIL: test_maximum_encoding_length_pair_input (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_maximum_encoding_length_pair_input) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 1253, in test_maximum_encoding_length_pair_input
    self.assertGreater(len(seq0_tokens), 2 + stride)
AssertionError: 0 not greater than 4

======================================================================
FAIL: test_maximum_encoding_length_single_input (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_maximum_encoding_length_single_input) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 1151, in test_maximum_encoding_length_single_input
    self.assertGreater(
AssertionError: 0 not greater than 4 : Issue with the testing sequence, please update it, it's too short

======================================================================
FAIL: test_pretrained_model_lists (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_pretrained_model_lists)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 1027, in test_pretrained_model_lists
    self.assertGreaterEqual(len(self.tokenizer_class.pretrained_vocab_files_map), 1)
AssertionError: 0 not greater than or equal to 1

======================================================================
FAIL: test_save_and_load_tokenizer (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_save_and_load_tokenizer) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 719, in test_save_and_load_tokenizer
    self.assertDictEqual(before_vocab, after_vocab)
AssertionError: {b'\x[115 chars]'': 11, b'\x0b': 12, b'\x0c': 13, b'\r': 14, b[919 chars]': 0} != {b'\x[115 chars]'': 10, b'\x0b': 11, b'\x0c': 12, b'\r': 13, b[918 chars]': 0}
Diff is 4489 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_save_and_load_tokenizer (tests.models.rwkv5.test_tokenization_rwkv5.RWKV5TokenizationTest.test_save_and_load_tokenizer) [Rwkv5Tokenizer]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bbuf/工作目录/RWKV/transformers/tests/test_tokenization_common.py", line 745, in test_save_and_load_tokenizer
    self.assertDictEqual(before_vocab, after_vocab)
AssertionError: {b'\x[115 chars]'': 11, b'\x0b': 12, b'\x0c': 13, b'\r': 14, b[983 chars] 103} != {b'\x[115 chars]'': 10, b'\x0b': 11, b'\x0c': 12, b'\r': 13, b[982 chars] 103}
Diff is 4562 characters long. Set self.maxDiff to None to see it.
```

reproduce command is `python3 -m unittest tests/models/rwkv5/test_tokenization_rwkv5.py` with the pr latest commit.

ArthurZucker commented:
Having a look right now 🤗

ArthurZucker commented:
Created #29095 with the correct tokenizer

BBuf commented:
> Created #29095 with the correct tokenizer. note that more tests should be added. Do you want me to fully take over to have all the reviews integrated as well?

Thank you. I would be extremely grateful if that were possible. I initially tried to add this model to HuggingFace out of interest, but it seems I underestimated the difficulty. Moreover, due to my own work, the progress on this PR has been very slow. I am very sorry for that.

ArthurZucker commented:
Alright I'll see what I can do asap

SmerkyG commented:
Please note that the implementations for _convert_token_to_id and _convert_id_to_token are implemented kind of backwards and should be more like:
```    def _convert_token_to_id(self, token):
        """Converts a token (str) in an id using the vocab."""
        return self.decoder.get(token.encode("utf-8"), self.unk_token_id)

    def _convert_id_to_token(self, index):
        """Converts an index (integer) in a token (str) using the vocab."""
        return self.encoder.get(index)```

ArthurZucker commented:
I'm back from holidays! 🥳 Gotta catch up everything and update this PR!

SmerkyG commented:
Nice! I have RWKV 6 ready as well!

On Sun, Mar 17, 2024, 3:57 PM Arthur ***@***.***> wrote:

> I'm back from holidays! 🥳 Gotta catch up everything and update this PR!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/26963#issuecomment-2002647017>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDK33UMNVLVHGKGN5E6VGTYYYNWVAVCNFSM6AAAAAA6JF7ZBCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMBSGY2DOMBRG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

BBuf commented:
> I'm back from holidays! 🥳 Gotta catch up everything and update this PR!

Thank you. If you need my assistance, you can call on me at any time. Additionally, I have used the modified `Rwkv5Tokenizer` at https://huggingface.co/RWKV/rwkv-5-world-1b5 and https://huggingface.co/RWKV/rwkv-5-world-3b, and it works well for both individual inference and  batch inference.

ArthurZucker commented:
@SmerkyG feel free to open a PR for v6 if you want based on this one!

SmerkyG commented:
Current use of use_cache seems like it may not be appropriate for RWKV? The state is a required element and cannot really be regenerated slowly, except by essentially using 'use_cache'.

I converted my latest v5 code to match your latest and did a ton of bugfixes such that it now works again in full CPU mode. Please see https://github.com/SmerkyG/RWKV-World-HF-Tokenizer/blob/main/rwkv_world_v5_model_batch/modeling_rwkv5.py to compare for details.

I will put the v6 code in similar shape. 

Also, please note that the CUDA kernel is not currently generating a proper state, so if you batch the initial context and then run with the non-CUDA kernel afterward (which we should really rename to nocuda from cpu) it will give bad results. But pure no-cuda works fine.

BBuf commented:
> Current use of use_cache seems like it may not be appropriate for RWKV? The state is a required element and cannot really be regenerated slowly, except by essentially using 'use_cache'.
> 
> I converted my latest v5 code to match your latest and did a ton of bugfixes such that it now works again in full CPU mode. Please see https://github.com/SmerkyG/RWKV-World-HF-Tokenizer/blob/main/rwkv_world_v5_model_batch/modeling_rwkv5.py to compare for details.
> 
> I will put the v6 code in similar shape to prepare for an eventual pull request.
> 
> Also, please note that the CUDA kernel is not currently generating a proper state, so if you batch the initial context and then run with the non-CUDA kernel afterward (which we should really rename to nocuda from cpu) it will give bad results. But pure no-cuda works fine.

是的。

> Current use of use_cache seems like it may not be appropriate for RWKV? The state is a required element and cannot really be regenerated slowly, except by essentially using 'use_cache'.
> 
> I converted my latest v5 code to match your latest and did a ton of bugfixes such that it now works again in full CPU mode. Please see https://github.com/SmerkyG/RWKV-World-HF-Tokenizer/blob/main/rwkv_world_v5_model_batch/modeling_rwkv5.py to compare for details.
> 
> I will put the v6 code in similar shape to prepare for an eventual pull request.
> 
> Also, please note that the CUDA kernel is not currently generating a proper state, so if you batch the initial context and then run with the non-CUDA kernel afterward (which we should really rename to nocuda from cpu) it will give bad results. But pure no-cuda works fine.

Could you please submit a fix PR (Pull Request) for the implementation [here](https://github.com/BBuf/RWKV-World-HF-Tokenizer/blob/main/rwkv5_model/modeling_rwkv5.py) ? I will review it. Currently, the use_cache indeed does not take effect.

SmerkyG commented:
@BBuf Hmm yours has changed a lot since I forked it but is now quite mismatched from @ArthurZucker 's version in this PR. Shouldn't we be moving towards a single unified version? I unfortunately don't have time to do a lot of work on this - I've only been able to do this much because I need it for RWKV6 paper related evals. But I'm happy to share my working v6 (you can already see it) so you can update yours, and I will do what I can to get v5 cuda working fully with state generation if possible tonight.

SmerkyG commented:
working on a PR for you after all :)
